src/sources/aws_s3/mod.rs

```diff
@@ -0,0 +1,468 @@
+    content_type: Option<&str>,
+    body: rusoto_s3::StreamingBody,
+) -> Box<dyn tokio::io::AsyncRead + Send + Unpin> {
+    use async_compression::tokio_02::bufread;
+
+    let r = tokio::io::BufReader::new(body.into_async_read());
+
+    let mut compression = compression;
+    if let Auto = compression {
+        compression =
+            determine_compression(key, content_encoding, content_type).unwrap_or(Compression::None);
+    };
+
+    use Compression::*;
+    match compression {
+        Auto => unreachable!(), // is mapped above
+        None => Box::new(r),
+        Gzip => Box::new(bufread::GzipDecoder::new(r)),
+        Zstd => Box::new(bufread::ZstdDecoder::new(r)),
+    }
+}
+
+/// try to determine the compression given the:
+/// * content-encoding
+/// * content-type
+/// * key name (for file extension)
+///
+/// It will use this information in this order
+fn determine_compression(
+    key: &str,
+    content_encoding: Option<&str>,
+    content_type: Option<&str>,
+) -> Option<Compression> {
+    content_encoding
+        .and_then(|e| content_encoding_to_compression(e))
+        .or_else(|| content_type.and_then(|t| content_type_to_compression(t)))
+        .or_else(|| object_key_to_compression(key))
+}
+
+fn content_encoding_to_compression(content_encoding: &str) -> Option<Compression> {
+    use Compression::*;
+    match content_encoding {
+        "gzip" => Some(Gzip),
+        "zstd" => Some(Zstd),
+        _ => Option::None,
+    }
+}
+
+fn content_type_to_compression(content_type: &str) -> Option<Compression> {
+    use Compression::*;
+    match content_type {
+        "application/gzip" | "application/x-gzip" => Some(Gzip),
+        "application/zstd" => Some(Zstd),
+        _ => Option::None,
+    }
+}
+
+fn object_key_to_compression(key: &str) -> Option<Compression> {
+    let extension = std::path::Path::new(key)
+        .extension()
+        .and_then(std::ffi::OsStr::to_str);
+
+    use Compression::*;
+    extension.and_then(|extension| match extension {
+        "gz" => Some(Gzip),
+        "zst" => Some(Zstd),
+        _ => Option::None,
+    })
+}
+
+mod test {
+    #[test]
+    fn determine_compression() {
+        use super::Compression;
+
+        let cases = vec![
+            ("out.log", Some("gzip"), None, Some(Compression::Gzip)),
+            (
+                "out.log",
+                None,
+                Some("application/gzip"),
+                Some(Compression::Gzip),
+            ),
+            ("out.log.gz", None, None, Some(Compression::Gzip)),
+            ("out.txt", None, None, None),
+        ];
+        for (key, content_encoding, content_type, expected) in cases {
+            assert_eq!(
+                super::determine_compression(key, content_encoding, content_type),
+                expected
+            );
+        }
+    }
+}
+
+#[cfg(feature = "aws-s3-integration-tests")]
+#[cfg(test)]
+mod integration_tests {
+    use super::{sqs, AwsS3Config, Compression, Strategy};
```

Comment:
I abstained from using `use super::*` here to see how it felt. I think I like the explicitness better.

Reply:
I would personally use `use super::*` in tests, especially since basically all the use statements below are redundant to what's already imported in the parent module.

But, it's a minor nit, so I'm fine leaving this experiment in.

Reply:
Yeah, I'm not 100% on it. Some of the imports are duplicated, but there is also a much smaller set of imports in the integration test and I found it easier to see where certain types were coming from rather than having to check first the `integration_tests` imports and then `super`'s imports.